orttraining/orttraining/training_ops/cuda/collective/nccl_kernels.cc

```diff
@@ -13,18 +13,25 @@ Status NcclAllReduce::ComputeInternal(OpKernelContext* context) const {
   cudaStream_t stream = nullptr;  // Default stream
   ncclComm_t comm = nccl_->Comm(group_type_);
 
-  size_t input_count = 0;
   const void* input_data = context->Input<Tensor>(0)->DataRaw();
   void* output_data = context->Output(0, context->Input<Tensor>(0)->Shape())->MutableDataRaw();
   MLDataType onnx_type = context->Input<Tensor>(0)->DataType();
+
+  // Although we assumed the memory address is contiguous for the input, ORT pads activation tensors to 64 bytes aligned  
+  // and initializers to 256 bytes aligned. There are tiny padding gaps in the contiguous buffer space. 
+  // We have to AllReduce on the entire buffer, including the padding space. 
+  const Tensor* last_tensor = context->Input<Tensor>(context->InputCount() - 1);
+  int8_t* end_address = (int8_t*)last_tensor->DataRaw() + last_tensor->SizeInBytes();
```

Comment:
casting away const here. consider `reinterpret_cast<const char*>` instead.
more info: https://en.cppreference.com/w/cpp/language/reinterpret_cast
lot of detail there, but i think `char` is one of the special types that can be used to treat something as a byte array

Reply:
Thanks for the reminder. I have address your comment for all 3 kernels in a upcoming PR. Will include you in the review.